Add support for multipart upload (#160)

richardkiene opened the pull request:
* Fixes #160, but omits listing multipart uploads

dekobon commented on storage/objects.go:
```diff
@@ -205,6 +256,48 @@ func (s *ObjectsClient) Delete(ctx context.Context, input *DeleteObjectInput) er
+	PartsDirectoryPath string
+}
+
+type GetMpuHeaders struct {
```
Why is the headers returned from `GetMpuOutput` limited to these two values?

richardkiene replied:
Those are the only values listed in the JSON response in the [documentation](https://github.com/joyent/manta-muskie/blob/master/docs/index.md#getmultipartupload-get-loginuploads0-fidstate). If there is documentation elsewhere that I can use to build a more complete list, that would be awesome.

dekobon replied:
Here is an example request/response:
```
GET /elijah.zupancic/uploads/00b/00bf2f27-6370-c70b-bb29-b43784ae7b93/state HTTP/1.1
accept-version: ~1.0
Accept: application/json, */*
x-request-id: d578d434-0ef6-11e9-ae86-3748230df2b0
Date: Thu, 3 Jan 2019 01:28:11 GMT
Authorization: Signature keyId="/elijah.zupancic/keys/46:26:9a:c2:ef:0f:07:b2:ce:83:66:82:4f:3b:f7:f7",algorithm="rsa-sha256",signature="R3OAO2WP30iEsuf0C4Eo+6SbltGnO1eK35KNLQF0FOd99MHoVyHJzoclaLRCpU+XNVLCkJafWS0gKI2ltNMJuCxoaAfM+ZpzkRzAl1xxQXuiRExLBRlqsEGgtgVDnSl9YvLULbVcU5Ilg6zuF1q3bYcfEQzzGFYZXeNxKFBIk/kNneZ7spGx0rWW7jpxi5o4dyvRcO/YGyCJpYRWQ6xYq7/Ui9213tVBcv3LdxfEbsH6/pxQEHvUQQR6pA+oUEKgZfrl70XVNUHqmdVguZH1gXp10clKLtTXdZCjtZ8Vgy5k90ZjhMngG3fbHrV4f9g31MWlK90mvCjTHVWlSOZt7g=="
x-http-signing-time-ns: 245654
Host: us-east.manta.joyent.com
Connection: Keep-Alive
User-Agent: Java-Manta-SDK/3.3.1-SNAPSHOT (Java/1.8.0_162/Oracle Corporation)
Accept-Encoding: gzip,deflate

HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 677
Content-MD5: WGukq69WNC54VWn2lDPROw==
Date: Thu, 03 Jan 2019 01:27:32 GMT
Server: Manta
x-request-id: d578d434-0ef6-11e9-ae86-3748230df2b0
x-response-time: 452
x-server-name: 562a0e29-5024-4482-8f6c-26e0d95c5a36
Connection: Keep-alive

{"id":"00bf2f27-6370-c70b-bb29-b43784ae7b93","state":"done","targetObject":"/elijah.zupancic/stor/java-manta-integration-tests/ac7c6be5-b653-424c-a058-34bd28c65750/EncryptedServerSideMultipartManagerIT/2d811cf8-3c63-4590-9da8-e1c741925fc0","headers":{"m-encrypt-key-id":"integration-test-key","m-encrypt-iv":"F4nimOYmIM9ZKYxdh1IKEg==","m-encrypt-metadata":"","m-encrypt-metadata-hmac":"ruUylu7tJ6W/jys1B16MdA==","m-encrypt-type":"client/1","m-encrypt-cipher":"AES128/CTR/NoPadding","m-encrypt-metadata-iv":"RlaOEJkmTTBq4ql+h/DhBg==","m-encrypt-hmac-type":"HmacMD5"},"numCopies":2,"creationTimeMs":1543790652523,"partsMD5Summary":"eBs5bzKeZdZvWgrPfiPO8w==","result":"committed"}
```

With formatted json:
```json
{
	"id": "00bf2f27-6370-c70b-bb29-b43784ae7b93",
	"state": "done",
	"targetObject": "/elijah.zupancic/stor/java-manta-integration-tests/ac7c6be5-b653-424c-a058-34bd28c65750/EncryptedServerSideMultipartManagerIT/2d811cf8-3c63-4590-9da8-e1c741925fc0",
	"headers": {
		"m-encrypt-key-id": "integration-test-key",
		"m-encrypt-iv": "F4nimOYmIM9ZKYxdh1IKEg==",
		"m-encrypt-metadata": "",
		"m-encrypt-metadata-hmac": "ruUylu7tJ6W/jys1B16MdA==",
		"m-encrypt-type": "client/1",
		"m-encrypt-cipher": "AES128/CTR/NoPadding",
		"m-encrypt-metadata-iv": "RlaOEJkmTTBq4ql+h/DhBg==",
		"m-encrypt-hmac-type": "HmacMD5"
	},
	"numCopies": 2,
	"creationTimeMs": 1543790652523,
	"partsMD5Summary": "eBs5bzKeZdZvWgrPfiPO8w==",
	"result": "committed"
}
```

richardkiene replied:
Again, I think it's best to keep to the documented response values rather than any given response test. Once the documentation gets an upgrade we can come back around and update this. Otherwise I fear we could end up adding values that are not supported broadly or that may accidentally be changed by a later author of the Manta MPU code.

dekobon commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+
+	return response, nil
+}
+
```
Are we missing an API function for `getMpuPart`?

richardkiene replied:
I don't see one defined in the RFD or in the Muskie documentation.

dekobon replied:
I believe this would be found in this table: https://github.com/joyent/rfd/blob/master/rfd/0065/README.md#http-api

Under: `HEAD /$ACCOUNT/uploads/[0-f]+/uuid/:partNum`

dekobon replied:
No prob.

dekobon commented on storage/objects.go:
```diff
@@ -205,6 +256,48 @@ func (s *ObjectsClient) Delete(ctx context.Context, input *DeleteObjectInput) er
+	Id string
+}
+
+type ListMpuPart struct {
```
Do we want to add all of the fields that are returned as part of an object listing like (content-type, modified time, etc)?

richardkiene replied:
So the only documentation I've seen on this is [here](https://github.com/joyent/rfd/blob/master/rfd/0065/README.md#5-list-parts) and the rest of this was derived from the behavior of the existing node client and a reading of the muskie code. Am I missing a more detailed list of what is returned in the response?

dekobon replied:
Here is what was returned from one of the requests that I did:

```json
{
  "name": "0",
  "etag": "09de2770-0684-40e1-82d5-c0f881c03a24",
  "size": 5242880,
  "type": "object",
  "contentType": "application/octet-stream",
  "contentMD5": "vfGT/owyBJ6/PjRl/A4hyQ==",
  "mtime": "2019-01-03T01:55:56.384Z",
  "durability": 2
}
```

richardkiene replied:
Do you know if that is stable?

richardkiene replied:
I think there is a broad issue with documentation for MPU from the Manta side of things. I'll create a new ticket to get all of that updated and I'll create an issue to update our response values for MPU calls once that has happened.

richardkiene commented on storage/objects.go:
```diff
@@ -205,6 +256,48 @@ func (s *ObjectsClient) Delete(ctx context.Context, input *DeleteObjectInput) er
+	ContentMd5    string `json:content-md5`
+}
+
+type GetMpuOutput struct {
```
Mine too, but The High Priest Rob Pike commands it so, lest ye be ridiculed for thou not writing idiomatic Golang.

richardkiene commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	if input.ForceInsert {
+		absDirName := _AbsCleanPath(path.Dir(string(absPath)))
+		exists, _ := checkDirectoryTreeExists(c, ctx, absDirName)
+		if !exists {
```
That's a really great point. Would you be okay with deferring that to another PR? The pattern being used here is inherited from existing code and would be a bit of an overhaul to refactor.

dekobon replied:
Which point are you referring to?

twhiteman commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	idLength := len(id)
+	prefixLen, err := strconv.Atoi(id[idLength-1 : idLength])
+	if err != nil {
+		return errors.Wrap(err, "unable to commit mpu")
```
Maybe tweak the error string "unable to commit mpu" to be "invalid mpu prefix length" ?

twhiteman replied:
Also - a comment on why this is here - I'd be confused if I hadn't heard you already talk about this special byte.

richardkiene replied:
👍

twhiteman commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	reqInput := client.RequestInput{
+		Method:  http.MethodPost,
+		Path:    "/" + c.client.AccountName + "/uploads",
+		Headers: &http.Header{},
```
should this be `Headers: headers` here - looks like your passing in a new empty headers object?

richardkiene replied:
Good catch!

twhiteman commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	reqInput := client.RequestNoEncodeInput{
+		Method:  http.MethodPut,
+		Path:    partPath,
+		Headers: &http.Header{},
```
Another case of `Headers: headers` here.

richardkiene replied:
👍

twhiteman commented on storage/objects.go:
```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	partNum := strconv.FormatUint(input.PartNum, 10)
+	prefixLen, err := strconv.Atoi(id[idLength-1 : idLength])
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to upload part")
```
Another case of "invalid mpu prefix length"

richardkiene replied:
👍

richardkiene commented:
@dekobon I created https://github.com/joyent/triton-go/issues/162 , https://github.com/joyent/triton-go/issues/163 , https://github.com/joyent/triton-go/issues/164 to address your points.

@twhiteman I think I've addressed all of your concerns in my latest push.

The failing travis-ci check is unrelated to this commit (it's an images issue) and I've run the test suite on my machine without issue. Furthermore I've verified that the MPU example code works when run against this PR.

Let me know if you have any more changes or if I can merge this in.

twhiteman commented:
@richardkiene yep, from the go-lang point of view I think your code is good to go 👍
